### PR TITLE
ipex intel extension for pytorch integration

### PR DESCRIPTION
**Pull request by sywangyi**
What does this PR do?
Implements feature request https://github.com/huggingface/accelerate/issues/700

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -1556,6 +1571,44 @@ def _prepare_megatron_lm(self, *args):
...
+        optimizer = None
+        dataloader = None
+        result = [obj for obj in args]
+        for obj in result:
+            if isinstance(obj, torch.utils.data.DataLoader) and dataloader is None:
+                dataloader = obj
+            elif isinstance(obj, torch.nn.Module):
+                model = obj
+            elif isinstance(obj, (torch.optim.Optimizer)):
+                optimizer = obj
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1593]: This assumes there is only one model/one optimizer and one dataloader. Many users pass along their training and validation dataloader together for instance, so there should be asserts and clear error message for those situations.

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -1556,6 +1571,44 @@ def _prepare_megatron_lm(self, *args):
...
+        if ipex_plugin.do_fusion and optimizer is None:
+            model.eval()
+            model = ipex.optimize(model, dtype=ipex_plugin.dtype, level="O1")
+            model = ipex_plugin.torch_jit_model_eval(model, dataloader, training=False)
+        else:
+            if not model.training:
+                model.train()
+            model, optimizer = ipex.optimize(
+                model, dtype=ipex_plugin.dtype, optimizer=optimizer, inplace=True, level="O1"
+            )
```
Comment (sgugger) [inline src/accelerate/accelerator.py:1603]: This assumes that the user will always pass along the optimizer with the model when training it, and that when they don't pass an optimizer, they are doing evaluation only. As said above, many users prepare everything together. Those assumptions should be clearly documented as they break from the rest of the library.

File: src/accelerate/utils/dataclasses.py [hunk 3]
```diff
@@ -1324,3 +1324,65 @@ def set_tensorboard_logging_options(self):
...
+            self.do_fusion = False
+
+    def set_mixed_precision(self, mixed_precision):
+        if mixed_precision == "fp16":
+            raise ValueError("Tried to use `fp16` but it is not supported on cpu")
+        elif mixed_precision == "bf16":
+            self.dtype = torch.bfloat16
+
+    def torch_jit_model_eval(self, model, dataloader, training=False):
+        if not training:
```
Comment (sgugger) [inline src/accelerate/utils/dataclasses.py:1357]: This is only used with `training=False` so there is no point having this argument and then this test.

**Review by sgugger — COMMENTED** [review]
Thanks for opening a fresh PR. I have some comments, and we would also need to add tests for this.

File: src/accelerate/accelerator.py [hunk 4]
```diff
@@ -1260,7 +1275,7 @@ def prepare_model(self, model: torch.nn.Module, device_placement=None):
         if self.distributed_type == DistributedType.TPU and self.state.fork_launched:
             model = xmp.MpModelWrapper(model).to(self.device)
         # torch.compile should be called last.
-        if self.state.dynamo_plugin.backend != DynamoBackend.NO:
+        if self.state.dynamo_plugin and self.state.dynamo_plugin.backend != DynamoBackend.NO:
```
[the remark this answers is not in the document]
Comment (sywangyi) [reply src/accelerate/accelerator.py]: fix it

File: src/accelerate/accelerator.py (hunk 2 above)
Comment (sywangyi) [reply src/accelerate/accelerator.py]: yes, we should not take the assumption like this. ipex will be only used for training part optimization as described above

**Comment by sywangyi** [comment]
> Thanks for opening a fresh PR. I have some comments, and we would also need to add tests for this.

I add testcase, please help review, later, I will add doc about this

**Review by sgugger — APPROVED** [review]
Thanks for your work on this! @pacman100 could you have a second look?

**Comment by sywangyi** [comment]
@pacman100 could you help with the review?

**Review by pacman100 — COMMENTED** [review]
Hello, Thank you @sywangyi for adding ipex support and iterating on the previous PR 🤗. I believe adding docs with the below details would help users to get a better idea about when and how to use it:
1. Brief intro on what ipex does and what is integrated currently in accelerate (only training for now)
2. How to use the current integration
3. Link to blog/original docs showing benefits when using ipex on CPU in comparison to without it

For reference, please find the DeepSpeed integration docs here: https://huggingface.co/docs/accelerate/usage_guides/deepspeed

**Comment by sywangyi** [comment]
agree, I will create a new PR for the document. For this PR, could you help with the merge?

**Review by pacman100 — APPROVED** [review]

**Pull request merged**